js/impress.js

```diff
@@ -668,6 +670,334 @@
 // will help somebody learn at least some part of it.
 
 /**
+ * Garbage collection utility
+ *
+ * This library allows plugins to add elements and event listeners they add to the DOM. The user
+ * can call `impress().lib.gc.teardown()` to cause all of them to be removed from DOM, so that
+ * the document is in the state it was before calling `impress().init()`.
+ *
+ * In addition to just adding elements and event listeners to the garbage collector, plugins
+ * can also register callback functions to do arbitrary cleanup upon teardown.
```

Comment:
If we're still goin with the `gc` pattern, can we add this into the [documentation](https://github.com/impress/impress.js/blob/525e4fc5f2c2da28323672ea7a41ac574475819f/DOCUMENTATION.md)?

Reply:
This, and next PRs from @henrikingo are going to introduce changes and new plugins.

Yes, in the end they should be properly documented. But we made a specific `dev` branch for merging these without releasing. So I'm fine with merging all the work first, trying it out as a whole and then finding the best way to document it - examples for plugins, API docs, etc.

Reply:
That's cool, it's just that we might lose track later due to the huge amount of changes. It could be useful to review each PR in isolation so that it's more "manageable" if you get what I mean...

Reply:
I know what you mean.

Currently this file has enough documentation in comments to understand what it does and how to use it. So it's enough to give it a review.

Once we have a larger scope (of all the libs and plugins) it may be easier to decide what and how to document.
It's quite likely that because of 'de-centralized' nature of plugins also documentation will not be just one file anymore and will live closer to the code of individual parts.

Reply:
> Currently this file has enough documentation in comments to understand what it does and how to use it. So it's enough to give it a review.

My point for an incremental PR is not to understand what the GC API does and how to use it. It's to understand how it's being used within a plugin by providing the minimum amount of code and API exposure to satisfy that plugin's requirement, nothing else. That allows us to improve the architecture on top of examples and proper context of how it's being used.

For example, we can start to merge plugins that don't need a lot of exposed APIs to work and expose those APIs as more PRs and plugins are submitted.

Does it makes sense?

Reply:
It makes sense in theory. And I would agree with such approach if we would be talking about new things being implemented.

But we are talking about a code that is already implemented, plugins that already rely on libraries, teardown, etc in their current form.
So it would require quite a lot of additional unnecessary work from @henrikingo to carve PRs down to only parts needed for a single PR and plugin.

I suppose this PR feels quite big, because it lays down libraries for all plugins that come later. It will be simpler later to see PRs of single plugins and their usage of libraries.
That's also why I mentioned couple of times that in this case it may make sense to wait with some changes until we have more stuff merged. Because it's not new stuff being implemented and easily changed - it's code already implemented that we want to merge and possibly improve once we have complete picture.

Reply:
> But we are talking about a code that is already implemented, plugins that already rely on libraries, teardown, etc in their current form.
So it would require quite a lot of additional unnecessary work from @henrikingo to carve PRs down to only parts needed for a single PR and plugin.

Cool, that makes sense.